### ISSUE TITLE
Speed up browser client

### ISSUE THREAD

**Issue by rubensworks** [comunica/comunica#438]
#### Issue type:

- :snail: Performance issue

____
#### Description:

Simple queries seem to be slower than they should be.

Comunica: http://query.linkeddatafragments.org/#transientDatasources=http%3A%2F%2Fdata.linkeddatafragments.org%2Fugent-biblio&query=SELECT%20%3Fname%20%3Fpublication%20WHERE%20%7B%0A%20%20%3Fauthor%20foaf%3Aname%20%3Fname.%0A%20%20%3Fpublication%20dc%3Acreator%20%3Fauthor.%0A%7D

Old LDF client: http://client.linkeddatafragments.org/#transientDatasources=http%3A%2F%2Fdata.linkeddatafragments.org%2Fugent-biblio&query=SELECT%20%3Fname%20%3Fpublication%20WHERE%20%7B%0A%20%20%3Fauthor%20foaf%3Aname%20%3Fname.%0A%20%20%3Fpublication%20dc%3Acreator%20%3Fauthor.%0A%7D

**Comment by DoctorBud** [comment]
An additional example of performance issues with Client.js vs Comunica...

The following query takes 3 to 4 times longer in Comunica vs the legacy `Client.js` library.

If I create `query.txt` with the following:

```sparql
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
PREFIX wdp: <http://www.wikidata.org/prop/direct/>
PREFIX wde: <http://www.wikidata.org/entity/>
PREFIX schema: <http://schema.org/>

SELECT DISTINCT ?beatle ?name ?pseudonym ?description ?image WHERE {
  wde:Q1299 wdp:P527 ?beatle .
  ?beatle rdfs:label ?name .
  ?beatle wdp:P742 ?pseudonym .
  ?beatle schema:description ?description .
  ?beatle wdp:P18 ?image .
}
```

then I see (comunica):

```bash
$ time NODE_ENV=production node ../comunica/packages/actor-init-sparql/bin/query.js https://query.wikidata.org/bigdata/ldf -f query.txt|wc -l
    7400

real	0m7.538s
user	0m5.109s
sys	0m0.339s
```

vs (old Client.js):

```bash
$ time NODE_ENV=production bin/ldf-client https://query.wikidata.org/bigdata/ldf -f query.txt|wc -l

    7400

real	0m1.846s
user	0m1.099s
sys	0m0.143s
```

So basically 1.85 sec vs 7.5 sec, even though both produce 7400 results.

**Comment by rubensworks** [comment]
Thanks for sharing this! This should definitely be optimizable.

**Comment by DoctorBud** [comment]
In trying to simplify the above, I came upon another bug/anomaly. If you create `query.txt` as follows:

```sparql
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
PREFIX wdp: <http://www.wikidata.org/prop/direct/>
PREFIX wde: <http://www.wikidata.org/entity/>

SELECT ?beatle ?name ?description ?image WHERE {
  wde:Q1299 wdp:P527 ?beatle .
  ?beatle rdfs:label ?name .
  ?beatle wdp:P18 ?image .
}
```

Then you still see a performance disparity, but you also see that the Comunica query produces 181 results, compared to 146 results for legacy Client.js.

Comunica:

```bash
$ time NODE_ENV=production node ../comunica/packages/actor-init-sparql/bin/query.js https://query.wikidata.org/bigdata/ldf -f query.txt|wc -l
     181

real	0m2.017s
user	0m1.120s
sys	0m0.124s
```

Legacy Client.js:

```bash
$ time NODE_ENV=production bin/ldf-client https://query.wikidata.org/bigdata/ldf -f query.txt|wc -l
     146

real	0m1.070s
user	0m0.331s
sys	0m0.051s
```

Note that if you add `DISTINCT` to the query:

```sparql
SELECT DISTINCT ?beatle ?name ?image WHERE {
```

then Comunica returns less results than legacy Client.js:

```bash
$ time NODE_ENV=production node ../comunica/packages/actor-init-sparql/bin/query.js https://query.wikidata.org/bigdata/ldf -f query.txt|wc -l
     121

real	0m1.921s
user	0m1.111s
sys	0m0.121s


$ time NODE_ENV=production bin/ldf-client https://query.wikidata.org/bigdata/ldf -f query.txt|wc -l
     137

real	0m0.992s
user	0m0.324s
sys	0m0.047s
```

Weird

**Comment by rubensworks** [comment]
@DoctorBud That's not good. I'll look into it.
I assume that's on the feature/metadata branch?

**Comment by DoctorBud** [comment]
@rubensworks Yes, `feature/metadata` branch. The master branch cannot perform this at all against the TPF server.

**Comment by rubensworks** [comment]
For reference, the significant performance difference has been fixed with #473.
Note that Comunica is still slightly slower, but the difference is only minor, and acceptable considering the additional things Comunica does.
I'll leave this issue open in any case for figuring out performance improvements.

**Comment by rubensworks** [comment]
Other overhead is probably caused by https://github.com/comunica/comunica/issues/561
